Sources/SwiftProtobuf/Message+AsyncMessageSequence.swift

```diff
@@ -0,0 +1,135 @@
+//
+// Copyright (c) 2023 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+///
+/// Extensions to `Message` to provide an async sequence over a binary delimited protobuf stream.
+///
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+@available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+extension Message {
+  
+  /// Creates an asynchronous sequence of messages decoded from resource bytes
+  public static func asyncSequence(
+    asyncBytes: URL.AsyncBytes,
+    extensions: ExtensionMap? = nil,
+    partial: Bool = false,
+    options: BinaryDecodingOptions = BinaryDecodingOptions()
+  ) -> AsyncMessageSequence<Self> {
+    let messageFactory = MessageFactory<Self>(
+      extensions: extensions,
+      partial: partial,
+      options: options
+    )
+    return AsyncMessageSequence(
+      asyncBytesIterator: asyncBytes.makeAsyncIterator(),
+      messageFactory: messageFactory
+    )
+  }
+}
+
+@available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+public struct AsyncMessageSequence<M: Message> : AsyncSequence, Sendable {
+  
+  let asyncBytesIterator: URL.AsyncBytes.AsyncIterator
+  fileprivate let messageFactory: MessageFactory<M>
+  
+  /// The message type in this asynchronous sequence.
+  public typealias Element = M
+  
+  /// An asynchronous iterator that produces the messages of this asynchronous sequence
+  public struct AsyncIterator : AsyncIteratorProtocol, Sendable {
+    
+    public var iter: URL.AsyncBytes.AsyncIterator
+    fileprivate let messageFactory: MessageFactory<M>
+    
+    /// Aysnchronously reads the next varint
+    @inlinable public mutating func nextVarInt() async throws -> UInt64? {
+      var messageSize: UInt64 = 0
+      var shift: UInt64 = 0
+      
+      while let byte = try await iter.next() {
+        messageSize |= UInt64(byte & 0x7f) << shift
+        shift += UInt64(7)
+        if shift > 35 {
+          throw BinaryDecodingError.malformedProtobuf
+        }
+        if (byte & 0x80 == 0) {
+          return messageSize
+        }
+      }
+      if (shift > 0) {
+        // The stream has ended inside a varint.
+        throw BinaryDecodingError.truncated
+      }
+      return nil // End of stream reached.
+    }
+    
+    /// Asynchronously advances to the next message and returns it, or ends the
+    /// sequence if there is no next message.
+    ///
+    /// - Returns: The next message, if it exists, or `nil` to signal the end of
+    ///   the sequence.
+    public mutating func next() async throws -> M? {
+      guard let messageSize = try await nextVarInt() else {
+        return nil
```

Comment:
Whenever we return early in this `next()` method we should make sure to `nil` out the `iterator` to make sure we never call `next()` on it again.

Reply:
Done.